pkgs/development/interpreters/php/generic.nix

```diff
@@ -170,19 +170,19 @@ let
               ln -s ${extraInit} $out/lib/php.ini
 
               if test -e $out/bin/php; then
-                wrapProgram $out/bin/php --set PHP_INI_SCAN_DIR $out/lib
+                wrapProgram $out/bin/php --set-default PHP_INI_SCAN_DIR $out/lib
```

Comment:
Previous attempt https://github.com/NixOS/nixpkgs/pull/157287.

IIRC, the issue with `--set-default` is that it would break other extensions.

Reply:
Yup, and that is the intended behavior. If I specify a custom php.UBI, I start with a blank slate

Reply:
Well, the Nixpkgs’s PHP infrastructure relies on this to enable even the default extensions like PDO or session so without this, it will not have almost any.

Specifically, when it [sets `PHP_INI_SCAN_DIR` to empty string](https://github.com/composer/xdebug-handler/blob/cf19e3380ffd84cea4a2cbfcc599ea07ef01f18a/src/XdebugHandler.php#L443), wrapper would [ignore the default value provided with `--set-default` ](https://github.com/NixOS/nixpkgs/blob/5911be300ddffbb55c9f7af2a06132c2de61acf1/pkgs/build-support/setup-hooks/make-wrapper.sh#L130).

But maybe we should not use `PHP_INI_SCAN_DIR` to control extensions and instead patch PHP to also respect `NIX_PHP_INI_SCAN_DIR` or `NIX_PHP_ENABLE_EXTENSION` which the packages like xdebug-handler would leave alone.

Reply:
Yup, that is exactly intended. If I provide a custom ini path it's my duty to load the extensions I want. Case in point, `composer/xdebug-handler` will replicate all but the xdebug config when using the persistent option

Reply:
But I am **not** providing custom ini path, Nixpkgs is doing it (transparently to the user).

Reply:
Just to make sure we are not talking past each other, let me restate my understanding of the situation.

The `php` executable in nix is a wrapper script that sets `PHP_INI_SCAN_DIR` so that nix can load a custom php.ini at a non-default location. That stock php.ini loads whatever extensions are available.

Because there are tools out there, that set `PHP_INI_SCAN_DIR` to a custom location. What those tools are attempting to achieve is to be able to load a completely different PHP configuration and make that persistent for all sub processes.

Now, the whole point is, that once a custom ini location is specified, all stock configuration from nix is gone and instead (not additionally!) the custom configuration is loaded. That’s exactly the same behavior as if nix would not set the stock nix configuration via `PHP_INI_SCAN_DIR` but instead use the built-in php.ini location.

```
php --info | grep -E "Additional .ini files|Configuration File \(php.ini\)"
Configuration File (php.ini) Path => /nix/store/s7rgnbhfbrbli6py2f726hd89ch54r73-php-8.2.3/lib
Additional .ini files parsed => /nix/store/vn7h4va52s0gbhys2chiwdkr3rxnzvz1-php-with-extensions-8.2.3/lib/php.ini
```

```
/nix/store/vn7h4va52s0gbhys2chiwdkr3rxnzvz1-php-with-extensions-8.2.3/bin/.php-wrapped --info | grep -E "Additional .ini files|Configuration File \(php.ini\)"
Configuration File (php.ini) Path => /nix/store/s7rgnbhfbrbli6py2f726hd89ch54r73-php-8.2.3/lib
Additional .ini files parsed => (none)
```

That’s why using `--set-default` is the right choice. It behaves symmetrically to `PHP_INI_SCAN_DIR` vs. built-in location and is exactly what any tool would expect that tinkers with `PHP_INI_SCAN_DIR`.

What `composer/xdebug-handler` does under the hood is to copy all of the current configuration except the settings that enable xdebug. Conceptually, one could implement functionality that does something similar but for opcache so that the current process is restarting without opcache. Or without PDO, or xmlreader or whatever. 

Point is: dropping all stock configuration is the right thing, as it’s the job of the tool building upon `PHP_INI_SCAN_DIR` to reconfigure PHP in whatever way desired.

Reply:
> That stock php.ini loads whatever extensions are available.

The issue there is no stock `php.ini`:

```ShellSession
$ php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /nix/store/9dr0yiky45hnrvgkk2gybijxbssm7b80-php-8.2.3/lib
Loaded Configuration File => (none)
Scan this dir for additional .ini files => /nix/store/b7nhlzic3y18ckngxg3zxjqnicjbqajd-php-with-extensions-8.2.3/lib
Additional .ini files parsed => /nix/store/b7nhlzic3y18ckngxg3zxjqnicjbqajd-php-with-extensions-8.2.3/lib/php.ini
```

Compare with Fedora:

```ShellSession
$ php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /etc
Loaded Configuration File => /etc/php.ini
Scan this dir for additional .ini files => /etc/php.d
Additional .ini files parsed => /etc/php.d/10-opcache.ini,
$ PHP_INI_SCAN_DIR= php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /etc
Loaded Configuration File => /etc/php.ini
Scan this dir for additional .ini files => (none)
Additional .ini files parsed => (none)
```

And since because

1. Nix does not allow dependency cycles between packages
2. There is no central location where PHP could look for extensions
3. Nixpkgs builds extensions as separate packages

Nixpkgs needs some method of passing the extensions to php. We have decided to use `PHP_INI_SCAN_DIR`.

> Now, the whole point is, that once a custom ini location is specified, all stock configuration from nix is gone and instead (not additionally!) the custom configuration is loaded. That’s exactly the same behavior as if nix would not set the stock nix configuration via `PHP_INI_SCAN_DIR` but instead use the built-in php.ini location.

Second issue, Nixpkgs makes this worse since it goes further in decoupling modules from core PHP package:

```ShellSession
$ PHP_INI_SCAN_DIR= php -m
[PHP Modules]
Core
date
hash
json
libxml
pcre
Phar
random
Reflection
SPL
standard
xml

[Zend Modules]
```

Again, contrast this with Fedora:

```ShellSession
$ PHP_INI_SCAN_DIR= php -m
[PHP Modules]
Core
date
filter
hash
json
libxml
openssl
pcntl
pcre
readline
Reflection
session
SPL
standard
zlib

[Zend Modules]
```

> Point is: dropping all stock configuration is the right thing, as it’s the job of the tool building upon `PHP_INI_SCAN_DIR` to reconfigure PHP in whatever way desired.

And third issue, if you drop the `ini` file from the wrapped package, you will not be able to re-add the extensions since there is no [`extension_dir`](https://www.php.net/manual/en/ini.core.php#ini.extension-dir) config.

So perhaps my point is that Nixpkgs’s wrapper is unsuitable for any use case that touches `PHP_INI_SCAN_DIR` and this PR just trades one breakage for another.

I guess an optimal solution would be patching PHP to accept `NIX_PHP_CONFIG_FILE` and then have the wrapper use that instead of relying on `PHP_INI_SCAN_DIR`. We could make `php.buildEnv` do a `symlinkJoin` of all enabled extensions and pass it through `extension_dir` config option in the wrapper.

Reply:
> What composer/xdebug-handler does under the hood is to copy all of the current configuration except the settings that enable xdebug.

I missed that xdebug-handler does this when skimming the code. If every major tool uses xdebug-handler or implements similar feature, it should not be such a problem then.

Since
- the [original change](https://github.com/NixOS/nixpkgs/pull/125405) have been reverted due to issues with duplicates but that should not happen with `--set-default`,
- the [previous attempt](https://github.com/NixOS/nixpkgs/pull/157287) just appears to have stalled (maybe due to lack of confidence), and
- the other issue I vaguely recalled was actually caused by the tool running `php` with `-n` flag

I retract my objection.